src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+            }
+        }
+
+        private static BsonDocument ToDocument(byte[] bytes)
+        {
+            MemoryStream stream = new MemoryStream(bytes);
+            using (var jsonReader = new BsonBinaryReader(stream))
+            {
+                var context = BsonDeserializationContext.CreateRoot(jsonReader);
+                return BsonDocumentSerializer.Instance.Deserialize(context);
+            }
+        }
+
+        private static byte[] ToBytes(BsonDocument doc)
+        {
+            BsonBinaryWriterSettings settings = new BsonBinaryWriterSettings()
+            {
+#pragma warning disable 618
+                GuidRepresentation = GuidRepresentation.Standard
+#pragma warning restore 618
+            };
+            return doc.ToBson(null, settings);
+        }
+
+        private class ClientFirst : ISaslStep
+        {
+            private readonly byte[] _bytesToSendToServer;
+            private readonly byte[] _nonce;
+            private readonly UsernamePasswordCredential _credential;
+            private readonly string _sessionToken;
+
+            public ClientFirst(
+                byte[] bytesToSendToServer,
+                byte[] nonce,
+                UsernamePasswordCredential credential,
+                string sessionToken)
+            {
+                _bytesToSendToServer = bytesToSendToServer;
+                _nonce = nonce;
+                _credential = credential;
+                _sessionToken = sessionToken;
+            }
+
+            public byte[] BytesToSendToServer
+            {
+                get { return _bytesToSendToServer; }
+            }
+
+            public bool IsComplete
+            {
+                get { return false; }
+            }
+
+            public ISaslStep Transition(SaslConversation conversation, byte[] bytesReceivedFromServer)
+            {
+                var serverFirstMessageDoc = ToDocument(bytesReceivedFromServer);
+                var serverNonce = serverFirstMessageDoc["s"].AsByteArray;
+                var host = serverFirstMessageDoc["h"].AsString;
+
+                if (serverNonce.Length != randomLength * 2 || !serverNonce.Take(randomLength).SequenceEqual(_nonce))
+                {
+                    throw new MongoAuthenticationException(conversation.ConnectionId, message: "Server sent an invalid nonce.");
+                }
+
+                var tuple = AwsSignatureVersion4.SignRequest(
+                    _credential.Username,
+                    _credential.GetInsecurePassword(),
+                    _sessionToken,
+                    serverNonce,
+                    host);
+
+                var document = new BsonDocument()
+                    .Add("a", tuple.Item1)
+                    .Add("d", tuple.Item2);
+
+                if (_sessionToken != null)
+                {
+                    document.Add("t", _sessionToken);
+                }
+
+                var clientSecondMessageBytes = ToBytes(document);
```

Comment:
NOTE: previously we used encoding.GetBytes for similar case https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Authentication/ScramShaAuthenticator.cs#L253 Not sure that document is better to use here. I will come back to it in my next round review.

Reply:
My point that previously we didn't use BsonDocument for that target. Instead we used `Utf8Encodings.Strict`. See here: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Authentication/ScramShaAuthenticator.cs#L253
But I don't think that there is anything wrong in the current implementation. I will rely on other reviewers' opinion here

Reply:
My apologies if I'm misunderstanding your comment Dima, but are you saying that we shouldn't use a BsonDocument here?

If so, I think the key difference between SCRAM-SHA and MONGODB-AWS is that the `payload` field in SCRAM-SHA uses the "SASL map" format (https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Authentication/ScramShaAuthenticator.cs#L210), which is essentially a fancy comma separated value string, and the payload field in MONGODB-AWS uses BSON (https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#mongodb-aws). Therefore, I believe the approach of using a BsonDocument here is correct.

Reply:
Why can't we just use:

    var clientSecondMessageBytes = document.ToBson();

Reply:
@vincentkam , I agree with you. Originally I thought that I saw comma separated data here, but probably i was wrong